### ISSUE TITLE
Unable to train or inference with GPU on Windows Machine

### ISSUE THREAD

**Issue by TomC868** [BVLC/caffe#3985]
I use to be able to train on my Windows machine with [happynear's Caffe port](https://github.com/happynear/caffe-windows) however, I moved to this repo and I believe there is some kind of configuration change of the Visual Studio Solution file that is causing me to encounter this issue now:
> syncedmem.cpp:56] Check failed: error == cudaSuccess (2 vs. 0) out of memory

The only change I made to the default CommonSettings.prop is I set up CudaArchitecture for a lower end GPUs: `<CudaArchitecture>computer_20,sm_20;computer_30,sm_30;</CudaArchitecture> `similar to how happynear setup his.

I guess the impulsive answer is that I am running out of GPU memory. However, I believe I should have enough GPU memory, as it works with happynear's repo, and nothing else is using GPU memory. Did anyone else encounter this?

**Comment by joelself** [comment]
Try putting this above the line that's failing to see if you really are allocating too much memory:
````cpp
    Caffe::DeviceQuery();
    LOG(INFO) << "Attempting to allocate " << size_ << " bytes of GPU memory." << std::endl;
````

**Comment by TomC868** [comment]
Thanks, that's a good idea. Here is the DeviceQuery() results:
```
 Device id:                     0
 Major revision number:         3
 Minor revision number:         0
 Name:                          Quadro K2100M
 Total global memory:           2147483648
 Total shared memory per block: 49152
 Total registers per block:     65536
 Warp size:                     32
 Maximum memory pitch:          2147483647
 Maximum threads per block:     1024
 Maximum dimension of block:    1024, 1024, 64
 Maximum dimension of grid:     2147483647, 65535, 65535
 Clock rate:                    666500
 Total constant memory:         65536
 Texture alignment:             512
 Concurrent copy and execution: Yes
 Number of multiprocessors:     3
 Kernel execution timeout:      Yes
```
And right before the error, it attempts to allocate just twenty million more bytes than it has:
`I0418 10:14:22.336494  1584 syncedmem.cpp:57] Attempting to allocate 2167603200 bytes of GPU memory.`

So the out of memory error makes sense here, but the interesting part is I looked at what Happynear's version needs for GPU memory (using the same dataset, prototxt files, and for the most part, the same source code) and it never even comes close to this, with the highest it tries to allocate being 44236800 bytes of memory. Here is a side by side of each allocation up until it crashes.

![caffe gpu memory usage](https://cloud.githubusercontent.com/assets/6855101/14610532/eff373cc-0553-11e6-9e25-4645d4fcb01d.png)

**Comment by TomC868** [comment]
As a second data point, I took this repository and built it on Linux using cmake and used the same prototxt and dataset that I was using on Windows, but the GPU memory used was vastly different than on Windows. 

![caffe gpu memory usage vs linux](https://cloud.githubusercontent.com/assets/6855101/14618681/6a543f78-0579-11e6-8a5a-b4fb320a1317.png)

In fact, the most the Linux build uses is 44236800 bytes, the same that HappyNear's repo did on Windows.

Thanks in advance for any ideas as to why this is.

**Comment by TomC868** [comment]
So I ran both programs through the debugger and found that HappyNear's version secretly uses CuDnn which I didn't have set up in this project. Because of that, my prototxt was using a BatchNorm layer, which calls the Forward_gpu() from conv_layer.cu instead of from cudnn_conv_layer.cu when using CuDnn. 

Thanks for the help.

**Comment by lasandcris** [comment]
Hey Tom, have you managed to fix the issue? I seem to have a very similar symptom when I run caffe.exe. Whenever that's executed there is a sudden crash. If you've found a solution to fix this could you share it. I disabled CuDnn and as an experiment and I did a caffe compile only for CPU too but it still persists. any help appreciated.

**Comment by TomC868** [comment]
My issue was that I was actually running out of GPU memory (hence the error == cudaSuccess (2 vs. 0) out of memory) but I was in denial about it because I use to be able to train on the GPU. Well it turned out that I was using CuDnn before without realizing it. CuDnn greatly optimizes GPU memory usage, which allows me to train on the GPU without running out of memory for my particular use cases.

If your crash persists on the CPU compiled only version of caffe, then it probably has to do with something else beside running out of GPU memory, as the CPU compiled version should be using your RAM instead. You might have to open up another issue with the exact message you get before it crashes to get some more insight.

**Comment by lasandcris** [comment]
Thanks for the reply. As it turns out I wasn't even explicitly running the CPU mode in caffe - and I think I messed up a few sys paths too - so I did a clean reinstall of everything. Caffe still crashed but I managed to figure out that the issue was my batch size. Now when I reduce my batch size caffe works perfectly but I do have to keep an eye on the base_lr, which needs to be increased by the factor of the reduction of the batch - or maybe even the sqrt of that (as I read somewhere else). So now the GPU works and I can train without any problem.
